### PR TITLE
prov/gni: Authorization key support

### PR DESCRIPTION
**Pull request by jswaro**
Added necessary code to support authorization keys for the GNI provider

Signed-off-by: James Swaro <jswaro@cray.com>

### CODE REVIEW

File: prov/gni/include/gnix.h
Comment (hppritcha) [inline prov/gni/include/gnix.h:636]: please add a description to the doxygen style comment

File: prov/gni/include/gnix.h
Comment (hppritcha) [inline prov/gni/include/gnix.h:671]: how does an auth_key get associated with a shared tx?  just curious

File: prov/gni/include/gnix_nic.h
Comment (hppritcha) [inline prov/gni/include/gnix_nic.h]: i'd rather the ```struct gnix_auth_key``` be added to the ``` gnix_nic_attr``` struct

File: prov/gni/src/gnix_dom.c
Comment (hppritcha) [inline prov/gni/src/gnix_dom.c]: we need something other than a hard-wired 256.

File: prov/gni/src/gnix_ep.c
Comment (hppritcha) [inline prov/gni/src/gnix_ep.c]: again please move auth_key into the nic attr struct.

File: prov/gni/src/gnix_init.c
Comment (hppritcha) [inline prov/gni/src/gnix_init.c]: what's going on here?  I don't understand the magical 8.

**Review by hppritcha — CHANGES_REQUESTED** [review]

File: prov/gni/src/gnix_init.c
Comment (jswaro) [reply prov/gni/src/gnix_init.c]: Runtime check to ensure the memory registration key is the size we expect it to be. It should be 64 bits, or the size of a uint64_t. 

Sloppiness on my part, but it needs be the same size as the returned value from fi_mr_key

File: prov/gni/src/gnix_dom.c
Comment (jswaro) [reply prov/gni/src/gnix_dom.c]: 256 or a define that says 256, because there is no point to having more than that if we only have 256 distinct ptags.

File: prov/gni/include/gnix_nic.h
Comment (jswaro) [reply prov/gni/include/gnix_nic.h]: ok

File: prov/gni/include/gnix.h
Comment (jswaro) [reply prov/gni/include/gnix.h:671]: This might be overstepping it on my end, but it made sense to me at the time because endpoints would have an auth key associated with them, but because a shared TX forces those endpoints to use a shared nic, all endpoints under that shared TX need to be using the same auth key.

File: prov/gni/include/gnix.h
Comment (jswaro) [reply prov/gni/include/gnix.h:636]: ok

**Comment by hppritcha** [comment]
using this PR as a guinea pig for a bit.
bot:lanl:retest

**Comment by hppritcha** [comment]
grrr...try again
bot:lanl:retest

**Comment by hppritcha** [comment]
gosh this is tedious, now cori complains.  try again.
bot:lanl:retest

**Comment by hppritcha** [comment]
okay almost there:
bot:lanl:retest

**Comment by hppritcha** [comment]
well this one doesn't look good.  let's try another pr.

**Comment by hppritcha** [comment]
bot:lanl:retest

**Comment by hppritcha** [comment]
turns out there were unexpected complications for jenkins.  on cori, there's no interposer, whereas on edison there is, and edison is running such an ancient distro that the script command doesn't have the options we need.  so built my own.  now testing.
bot:lanl:retest

**Comment by hppritcha** [comment]
hmmm... this PR needs some work.  It doesn't currently pass gnitest.  There are numerous tests which fail with this kind of signature
```
[[0;34mRUN [0m] wait_signal::has_data
boxfort-worker: prov/gni/src/gnix_fabric.c:549: _gnix_ep_getinfo: Assertion `0' failed.
[[0;34m----[0m] [0;1mWarning! The test `wait_signal::has_data` crashed during its setup or teardown.[0m
[[0;34m====[0m] Running [0;34m1[0m test from [0;33mwait_spin_adjust[0m:
[[0;34mRUN [0m] wait_spin_adjust::set_val
```

**Comment by jswaro** [comment]
Howard, that assertion is unrelated to my changes. That assertion is raised from the provider not understanding the mr_mode. Since authorization keys do not relate to mr modes (at this point), I recommend that it gets fixed outside of the scope of this issue.

**Comment by jswaro** [comment]
Ahh nevermind .... it appears my rebase re-introduced the default condition.

**Review by hppritcha — APPROVED** [review]

**Review by sungeunchoi — APPROVED** [review]
I didn't have a chance to do a through second review, but I think we should get this in asap.  The merge conflicts need to be resolved first tho.

**Pull request merged**
